### PR TITLE
Rhinoceros_Engine: add convert methods for Extrusion

### LINKED ISSUE #182: Rhinoceros_Toolkit: Add support for converts of Extrusions
#### Description:

Should be able to convert to and from Extrusions.

I know some initial studies has been done for this before with no success, but raising again to keep in mind.

### PR DESCRIPTION
**Pull request by alelom**
### Issues addressed by this PR

Closes #182

### Test files

Create any BHoM Extrusion, then try to convert it ToRhino(), and then FromRhino().

To test the FromRhino() from a script, you may want to temporarily modify the IFromRhino method as below:
<details>

```cs
public static BHG.IGeometry IFromRhino(this object geometry)	     
{	 
            // Code to force testing of Extrusion: first convert toRhino, then fromRhino.	
            if (geometry is BH.oM.Geometry.Extrusion)	
                return Convert.FromRhino((geometry as BH.oM.Geometry.Extrusion).ToRhino());	

            return (geometry == null) ? null : Convert.FromRhino(geometry as dynamic);
}
```

</details>

otherwise, due to the Goo casting being enforced for BHoM components, you'll never be able to pass a Rhino object to FromRhino() using Grasshopper components.

### CODE REVIEW

File: Rhinoceros_Engine/Convert/ToRhino.cs [hunk 1]
```diff
@@ -433,11 +433,29 @@ public static RHG.Brep ToRhino(this List<BHG.ISurface> surfaces)
...
+
+            var planarCurve = extrusion.Curve.IToRhino();
+
+            RHG.Plane curvePlane;
+            planarCurve.TryGetPlane(out curvePlane);
+
+
+            double extrHeight = extrusion.Direction.Length();
+
+            double angle = RHG.Vector3d.VectorAngle(curvePlane.Normal, extrusion.Direction.ToRhino()) * (180 / Math.PI);
```
Comment (IsakNaslundBh) [inline Rhinoceros_Engine/Convert/ToRhino.cs:451]: Minor comment, but why not keep it at radians? 😄

File: Rhinoceros_Engine/Convert/ToRhino.cs [hunk 2]
```diff
@@ -433,11 +433,29 @@ public static RHG.Brep ToRhino(this List<BHG.ISurface> surfaces)
...
+
+            if (angle > 90)
```
Comment (IsakNaslundBh) [inline Rhinoceros_Engine/Convert/ToRhino.cs:453]: Maybe should add a check if the direction and the curve are non-parallel within some tolerance? If not, maybe return null? Potentially to use the IsParallel from geometry engine maybe?

**Review by IsakNaslundBh — COMMENTED** [review]
Small comments below, have just looked here on GitHub, so not tested. Will be nice to get this in!

File: Rhinoceros_Engine/Convert/ToRhino.cs [hunk 3]
```diff
@@ -433,11 +433,29 @@ public static RHG.Brep ToRhino(this List<BHG.ISurface> surfaces)
...
+                extrHeight = -extrHeight; 
+
+            RHG.Extrusion extr = Rhino.Geometry.Extrusion.Create(planarCurve, extrHeight, extrusion.Capped);
+
+            return extr;
         }
```
Comment (rolyhudson) [inline Rhinoceros_Engine/Convert/ToRhino.cs:460]: Just having a quick look at this. Need to bear in mind that `BH.oM.Geometry.Extrusion` does not enforce extrusions perpendicular to the curve. So this outcome below possible with the current code. It is not clear to me if a BHoM extrusion is / was supposed to be constrained to only extrude normal to the plane of the curve.
Screenshot 2021-02-23 092022

Comment (alelom) [reply Rhinoceros_Engine/Convert/ToRhino.cs:487]: Thanks @rolyhudson . See commit https://github.com/BHoM/Rhinoceros_Toolkit/pull/204/commits/d24289bd6b966d4ac4e988c3fb8c079c8a04e2a5

**Comment by alelom** [comment]
@rolyhudson @al-fisher 
Because Rhino is enforcing perpendicularity btw base curve and extrusion direction, while BHoM gives more freedom, [I quickly added a feature to return a Sweep in case that condition is not met](https://github.com/BHoM/Rhinoceros_Toolkit/pull/204/commits/d24289bd6b966d4ac4e988c3fb8c079c8a04e2a5).

image

The method works by extrapolating the sweep rail from the centroid of the base curve (or middle point, if it is open). 

For some reason, the sweep is interrupted a bit before the top of the provided direction line, see below. I guess this feature can be improvable, but I'd rather raise a separate issue on this than keeping this PR open, which is already useful as it is for our applications.

image

File: Rhinoceros_Engine/Convert/ToRhino.cs (hunk 2 above)
Comment (alelom) [reply Rhinoceros_Engine/Convert/ToRhino.cs]: Done in https://github.com/BHoM/Rhinoceros_Toolkit/pull/204/commits/d24289bd6b966d4ac4e988c3fb8c079c8a04e2a5

File: Rhinoceros_Engine/Convert/ToRhino.cs (hunk 1 above)
Comment (alelom) [reply Rhinoceros_Engine/Convert/ToRhino.cs]: Changed in https://github.com/BHoM/Rhinoceros_Toolkit/pull/204/commits/d24289bd6b966d4ac4e988c3fb8c079c8a04e2a5

**Comment by al-fisher** [comment]
Good approach returning sweep for not perpendicular BHoM.Extrusions. Agreed.
Also agree on current limitation on Sweep cut off - worth raising separate issue on that

Standard extrude working well for me now in preview
Although can not recreate your skewed Sweep behaviour? Have you missing commits? Or have I messed up my end? @alelom

**Comment by alelom** [comment]
> Although can not recreate your skewed Sweep behaviour? Have you missing commits? Or have I messed up my end? @alelom
Sorry yes I was missing to save the file then commit!

**Review by al-fisher — DISMISSED** [review]
LGTM now thanks @alelom.

Standard Extrude working as expected - and can recreate your behaviour (including the potential endpoint/cut off issue) with the Rhino.Sweep/BHoM.NonOrthognalExtrude

If @rolyhudson and @IsakNaslundBh are happy - good to merge from me

**Comment by alelom** [comment]
Thanks @al-fisher , I had to push https://github.com/BHoM/Rhinoceros_Toolkit/pull/204/commits/4acbbf09fcbcaec2069e50d392c663704dfeba8a to fix build not working for Grasshopper as highlighted by CheckInstaller.

**Review by rolyhudson — APPROVED** [review]
LGTM. Issue with the interrupted sweep seems to be only when a circle. Here is an arc swept next to the manually modelled rhino version.
Screenshot 2021-02-23 154504
And rectangle with its direction line baked and selected.
Screenshot 2021-02-23 161403

**Pull request merged**